presto-main/src/main/java/com/facebook/presto/execution/scheduler/group/DynamicBucketNodeMap.java

```diff
@@ -63,16 +77,25 @@ public int getBucketCount()
     }
 
     @Override
-    public void assignOrUpdateBucketToNode(int bucketedId, InternalNode node)
+    public void assignOrUpdateBucketToNode(int bucketedId, InternalNode node, boolean cacheable)
     {
         checkArgument(bucketedId >= 0 && bucketedId < bucketCount);
         requireNonNull(node, "node is null");
-        bucketToNode.put(bucketedId, node);
+        bucketToNodeAndCache.put(bucketedId, new InternalNodeAndCache(node, cacheable));
     }
 
     @Override
     public boolean isDynamic()
     {
         return true;
     }
+
+    @Override
+    public Optional<List<InternalNode>> getBucketToNode()
+    {
+        if (bucketToNodeAndCache.size() == 0) {
```

Comment:
I'm not entirely sure if this is a good predicate since this would cover both of the following cases:
- There is initial preferred mapping
- There is no initial preferred mapping, but with at least one bucketToNode assignment by calling `assignOrUpdateBucketToNode`

Based on my understanding on the comment, the second case might not be desired?

Reply:
Yes you're right, both situation could happen.

By saying "the second case might not be desired",  I'm guessing you're referring to the call in `SectionExecutionFactory`?
if it's the call in `SectionExecutionFactory`, I think we should be safe, because we call `getBucketToNode` right after `nodePartitioningManager.getBucketNodeMap`, there is no situation for `assignOrUpdateBucketToNode` to happen.

But I understand your concern, maybe your suggestion is to have a different method name to indicate that it's initialized?

Reply:
Yeah I think if what we want to check is initial preferred mapping exists, let's make it exclusive and explicit :D

Reply:
Sounds reasonable! Will do